### PR TITLE
add firstName and lastName to UserDTO and IUser

### PR DESCRIPTION
**Pull request by VergilSkye**
As discussed in #14164. firstName and lastName were removed from the user for ~~security~~ privacy reasons, so they cannot be used as a display fields.

this pr returns the old behavior and allows firstName and lastName to be used, but I don't think it should be merged without discussing options.

 in my view, we have 3 options

1) Don't merge this pr and and classify this as custom behavior. _Focus on ~~security~~ privacy_

2) Merge this pr and rollback the old behavior. Old jdls, which use these fields, can be imported with as little friction as possible. _Focus on backward compatibility_

3) Allow jdl using firstName and lastName but dont change User and IUser. this will break the code but we can create a tip on how to make the code work again write custom code ( how to change User java, integration test and IUser class). Balance between 1 and 2. 

---

Please make sure the below checklist is followed for Pull Requests.

- [ ] [All continuous integration tests](https://github.com/jhipster/generator-jhipster/actions) are green
- [ ] Tests are added where necessary
- [ ] [jhipster-online](https://github.com/jhipster/jhipster-online) was updated if necessary
- [ ] Documentation is added/updated where necessary
- [ ] Coding Rules & Commit Guidelines as per our [CONTRIBUTING.md document](https://github.com/jhipster/generator-jhipster/blob/main/CONTRIBUTING.md) are followed

When you are still working on the PR, consider converting it to Draft (bellow reviewers) and adding `skip-ci` label, you can still see CI build result at your branch.

### CODE REVIEW

File: generators/server/templates/src/main/java/package/service/dto/UserDTO.java.ejs [hunk 1]
```diff
@@ -63,12 +69,30 @@ public class <%= asDto('User') %> {
...
+    }
+
+    public String getLastName() {
+        return lastName;
+    }
+
+    public void setLastName(String lastName) {
+        this.lastName = lastName;
+    }
+
```
Comment (mshima) [inline generators/server/templates/src/main/java/package/service/dto/UserDTO.java.ejs:87]:
```suggestion
<%_ for (field of _.sortedUniqBy(user.otherRelationships.map(relationship => relationship.relatedField), field => field.fieldName)) { _%>
    public String get<%= javaBeanCase(field.fieldName) %>() {
        return <%= field.fieldName %>;
    }

    public void set<%= javaBeanCase(field.fieldName) %>(String <%= field.fieldName %>) {
        this.<%= field.fieldName %> = <%= field.fieldName %>;
    }

<%_ } _%>
```

File: generators/server/templates/src/main/java/package/service/dto/UserDTO.java.ejs [hunk 2]
```diff
@@ -35,15 +35,21 @@ public class <%= asDto('User') %> {
 
     private String login;
 
+    private String firstName;
+
+    private String lastName;
+
```
Comment (mshima) [inline generators/server/templates/src/main/java/package/service/dto/UserDTO.java.ejs:41]:
```suggestion
<%_ for (field of _.sortedUniqBy(user.otherRelationships.map(relationship => relationship.relatedField), field => field.fieldName)) { _%>
    private String <%= field.fieldName %>;

<%_ } _%>
```

File: generators/server/templates/src/main/java/package/service/dto/UserDTO.java.ejs [hunk 3]
```diff
@@ -35,15 +35,21 @@ public class <%= asDto('User') %> {
...
         // Empty constructor needed for Jackson.
     }
 
     public <%= asDto('User') %>(<%= asEntity('User') %> user) {
         this.id = user.getId();
-        // Customize it here if you need firstName/lastName/etc
+        // Customize it here if you need, or not, firstName/lastName/etc 
         this.login = user.getLogin();
+        this.firstName = user.getFirstName();
+        this.lastName = user.getLastName();
```
Comment (mshima) [inline generators/server/templates/src/main/java/package/service/dto/UserDTO.java.ejs:52]:
```suggestion
<%_ for (field of _.sortedUniqBy(user.otherRelationships.map(relationship => relationship.relatedField), field => field.fieldName)) { _%>
        this.<%= field.fieldName %> = user.get<%= javaBeanCase(field.fieldName) %>();
<%_ } _%>
```

File: generators/server/templates/src/main/java/package/service/dto/UserDTO.java.ejs [hunk 4]
```diff
@@ -63,12 +69,30 @@ public class <%= asDto('User') %> {
...
+    }
+
     // prettier-ignore
     @Override
     public String toString() {
         return "<%= asDto('User') %>{" +
             "id='" + id + '\'' +
             ", login='" + login + '\'' +
+            ", firstName='" + firstName + '\'' +
+            ", lastName='" + lastName + '\'' +
```
Comment (mshima) [inline generators/server/templates/src/main/java/package/service/dto/UserDTO.java.ejs:95]:
```suggestion
<%_ for (field of _.sortedUniqBy(user.otherRelationships.map(relationship => relationship.relatedField), field => field.fieldName)) { _%>
            ", <%= field.fieldName %>='" + <%= field.fieldName %> + '\'' +
<%_ } _%>
```

File: generators/server/templates/src/test/java/package/web/rest/PublicUserResourceIT.java.ejs [hunk 5]
```diff
@@ -194,6 +200,8 @@ class PublicUserResourceIT <% if (databaseType === 'cassandra') { %>extends Abst
             .returnResult(<%= asDto('User') %>.class).getResponseBody().blockFirst();
 
         assertThat(foundUser.getLogin()).isEqualTo(DEFAULT_LOGIN);
+        assertThat(foundUser.getFirstName()).isEqualTo(DEFAULT_FIRSTNAME);
+        assertThat(foundUser.getLastName()).isEqualTo(DEFAULT_LASTNAME);
```
Comment (mshima) [inline generators/server/templates/src/test/java/package/web/rest/PublicUserResourceIT.java.ejs:204]:
```suggestion
<%_ for (field of _.sortedUniqBy(user.otherRelationships.map(relationship => relationship.relatedField), field => field.fieldName)) { _%>
        assertThat(foundUser.get<%= javaBeanCase(field.fieldName) %>()).isEqualTo(DEFAULT_<%= field.fieldName.toUpperCase() %>);
<%_ } _%>
```

File: generators/server/templates/src/test/java/package/web/rest/PublicUserResourceIT.java.ejs [hunk 6]
```diff
@@ -180,6 +184,8 @@ class PublicUserResourceIT <% if (databaseType === 'cassandra') { %>extends Abst
             .andExpect(status().isOk())
             .andExpect(content().contentType(MediaType.APPLICATION_JSON_VALUE))
             .andExpect(jsonPath("$.[*].login").value(hasItem(DEFAULT_LOGIN)))
+            .andExpect(jsonPath("$.[*].firstName").value(hasItem(DEFAULT_FIRSTNAME)))
+            .andExpect(jsonPath("$.[*].lastName").value(hasItem(DEFAULT_LASTNAME)))
```
Comment (mshima) [inline generators/server/templates/src/test/java/package/web/rest/PublicUserResourceIT.java.ejs:188]:
```suggestion
<%_ for (field of _.sortedUniqBy(user.otherRelationships.map(relationship => relationship.relatedField), field => field.fieldName)) { _%>
            .andExpect(jsonPath("$.[*].<%= field.fieldName %>").value(hasItem(DEFAULT_<%= field.fieldName.toUpperCase() %>)))
<%_ } _%>
```

File: generators/server/templates/src/test/java/package/web/rest/PublicUserResourceIT.java.ejs [hunk 7]
```diff
@@ -113,6 +113,10 @@ class PublicUserResourceIT <% if (databaseType === 'cassandra') { %>extends Abst
 
     private static final String DEFAULT_LOGIN = "johndoe";
 
+    private static final String DEFAULT_FIRSTNAME = "john";
+
+    private static final String DEFAULT_LASTNAME = "doe";
+
```
Comment (mshima) [inline generators/server/templates/src/test/java/package/web/rest/PublicUserResourceIT.java.ejs:119]:
```suggestion
<%_ for (field of _.sortedUniqBy(user.otherRelationships.map(relationship => relationship.relatedField), field => field.fieldName)) { _%>
    private static final String DEFAULT_<%= field.fieldName.toUpperCase %> = "<%= field.fieldName %>";

<%_ } _%>
```

**Review by mshima — CHANGES_REQUESTED** [review]
Additional fields should not be generated by default but on demand.

**Comment by VergilSkye** [comment]
Thanks for the reply and I added all changes @mshima but the PublicUserResourceIT  it's crashing because the _DEFAULT_FIRSTNAME_ and _DEFAUL_LASTNAME_ has wrong constants.
the following code:
```
 <%_ for (field of _.sortedUniqBy(user.otherRelationships.map(relationship => relationship.relatedField), field => field.fieldName)) { _%>
        private static final String DEFAULT_<%= field.fieldName.toUpperCase() %> = "<%= field.fieldName %>";
    <%_ } _%>
```
generates:
  private static final String DEFAULT_FIRSTNAME = "**firstName**";
  private static final String DEFAULT_LASTNAME  = "**lastName**";


but should be:
  private static final String DEFAULT_FIRSTNAME = "**john**";
  private static final String DEFAULT_FIRSTNAME = "**doe**";


Should i adding some function to convertToConstants ?
where it can be added?

File: generators/server/templates/src/test/java/package/web/rest/PublicUserResourceIT.java.ejs (hunk 6 above)
Comment (VergilSkye) [reply generators/server/templates/src/test/java/package/web/rest/PublicUserResourceIT.java.ejs]: this code this code's making IT crash, by generate wrong constatns, for firstName should be **john** and for lastName should be **doe**

File: generators/server/templates/src/test/java/package/web/rest/PublicUserResourceIT.java.ejs (hunk 7 above)
Comment (VergilSkye) [reply generators/server/templates/src/test/java/package/web/rest/PublicUserResourceIT.java.ejs]: this code this code's making IT crash, by generate wrong constatns, for firstName should be john and for lastName should be doe

File: generators/client/templates/angular/src/main/webapp/app/entities/user/user.model.ts.ejs [hunk 8]
```diff
@@ -22,12 +22,18 @@ _%>
 export interface IUser {
   id?: <%= idType %>;
   login?: string;
+  <%_ for (field of _.sortedUniqBy(user.otherRelationships.map(relationship => relationship.relatedField), field => field.fieldName)) { _%>
+    <%= field.fieldName %>?: string; 
+  <%_ } _%>
```
Comment (mshima) [inline generators/client/templates/angular/src/main/webapp/app/entities/user/user.model.ts.ejs:27]:
```suggestion
  <%_ for (field of user.fields.filter(field => !field.builtIn && field.relatedByOtherEntity)) { _%>
    <%= field.fieldName %>?: string; 
  <%_ } _%> 
```

File: generators/server/templates/src/main/java/package/service/dto/UserDTO.java.ejs [hunk 9]
```diff
@@ -35,15 +35,22 @@ public class <%= asDto('User') %> {
 
     private String login;
 
+    <%_ for (field of _.sortedUniqBy(user.otherRelationships.map(relationship => relationship.relatedField), field => field.fieldName)) { _%>
+        private String <%= field.fieldName %>;
+    <%_ } _%>
+
```
Comment (mshima) [inline generators/server/templates/src/main/java/package/service/dto/UserDTO.java.ejs:41]:
```suggestion
<%_ for (field of user.fields.filter(field => !field.builtIn && field.relatedByOtherEntity)) { _%>
        private String <%= field.fieldName %>;
<%_ } _%>

```

File: generators/server/templates/src/main/java/package/service/dto/UserDTO.java.ejs [hunk 10]
```diff
@@ -35,15 +35,22 @@ public class <%= asDto('User') %> {
...
     }
 
     public <%= asDto('User') %>(<%= asEntity('User') %> user) {
         this.id = user.getId();
-        // Customize it here if you need firstName/lastName/etc
+        // Customize it here if you need, or not, firstName/lastName/etc 
         this.login = user.getLogin();
+        <%_ for (field of _.sortedUniqBy(user.otherRelationships.map(relationship => relationship.relatedField), field => field.fieldName)) { _%>
+        this.<%= field.fieldName %> = user.get<%= javaBeanCase(field.fieldName) %>();
+        <%_ } _%>
```
Comment (mshima) [inline generators/server/templates/src/main/java/package/service/dto/UserDTO.java.ejs:53]:
```suggestion
<%_ for (field of user.fields.filter(field => !field.builtIn && field.relatedByOtherEntity)) { _%>
        this.<%= field.fieldName %> = user.get<%= javaBeanCase(field.fieldName) %>();
<%_ } _%>
```

**Review by mshima — CHANGES_REQUESTED** [review]

**Review by mshima — APPROVED** [review]
There is no test case that uses those custom fields, not sure is a feature we must add tests for.
I am supposing it's working.

**Comment by VergilSkye** [comment]
yeah, i don't know the jhipster policy for tests, but i already know a edge case.

if a end user create project and trying to import jdl that uses firtsName as display field, the jdl will be generate but the code will be broken. The end user need to regenerate the entity User.
I use this command to regenerate User ``` jhipster --force --with-entities```

**Comment by mshima** [comment]
With this PR it's the developer choice what fields to expose. It's not a privacy problem.

**Comment by mshima** [comment]
Thanks @VergilSkye

**Comment by VergilSkye** [comment]
Thanks @mshima  for your help, guidance and all the work done on the Jhipster project.

I am very excited about V7 and for having a small participation in this release.
